File: viz/lib/prov.ts
// Reader for one trainer run's W3C PROV-JSON document (the sidecar the trainer
// publishes per run — see training/prov.py, which is the sole producer). The
// snapshot points at the public copy through provenance.prov_ref; the models
// page fetches that URL and hands the parsed body here.
//
// This module is PURE: raw document in, a flat derivation chain out, no fetch
// and no clock. It exists so the parse is unit-testable against a document the
// emitter builds, rather than a hand-written one.
//
// FACTS-ONLY, mirroring the emitter's grounding rule: we surface exactly the
// nodes and edges the document carries. A run whose GTFS fetch failed publishes
// no feed entity, so `feed` is null and no feed node is drawn — never a
// fabricated "unknown feed". A document we cannot make sense of at all (not an
// object, or no trainer-run activity with grounded timestamps) parses to null,
// which the caller renders as "provenance unavailable" without touching the
// rest of the page.

export interface ProvRun {
  // The model version stamp, read off the run activity's id (mmly:run/<n>) —
  // the same integer a params key and the snapshot's params.trained_at carry.
  trainedAt: number;
  // Epoch seconds, from the activity's start/end xsd:dateTime literals.
  startedAt: number;
  endedAt: number;
}

export interface ProvAgent {
  codeSha: string;
  // null when the document did not record a dirty flag (the emitter omits it
  // when the trainer build couldn't tell) — distinct from a recorded `false`.
  dirty: boolean | null;
  producer: string | null;
}

export interface ProvFeed {
  version: string;
  sha256: string;
  start: string | null;
  end: string | null;
}

export interface ProvManifest {
  blake3: string;
  nAlertKeys: number;
  nVehicleKeys: number;
  manifestVersion: number;
}

export interface ProvArtifact {
  name: string;
  bucketKey: string;
  // Reconstructed from the document's wasDerivedFrom edges: this artifact was
  // built from the GTFS feed and/or the input manifest.
  derivedFromFeed: boolean;
  derivedFromManifest: boolean;
}

// The whole derivation chain, flattened for rendering: GTFS feed -> trainer run
// (its software agent) -> input manifest -> published artifacts. The snapshot
// the reader is looking at is appended by the caller, which holds it — it is
// not part of the run's own document.
export interface ProvChain {
  run: ProvRun;
  agent: ProvAgent | null;
  feed: ProvFeed | null;
  manifest: ProvManifest | null;
  artifacts: ProvArtifact[];
}

// A PROV-JSON element map (entity/activity/agent) as plain records, dropping any
// non-object member. A missing or wrong-shaped section is silence, not a parse
// failure, so it returns an empty map.
function recordSection(doc: Record<string, unknown>, key: string): Record<string, Record<string, unknown>> {
  const section = doc[key];
  if (!section || typeof section !== "object" || Array.isArray(section)) return {};
  const out: Record<string, Record<string, unknown>> = {};
  for (const [id, node] of Object.entries(section as Record<string, unknown>)) {
    if (node && typeof node === "object" && !Array.isArray(node)) {
      out[id] = node as Record<string, unknown>;
    }
  }
  return out;
}

// PROV-JSON carries typed values as { "$": value, "type": "xsd:…" } and plain
// values bare; unwrap both to the underlying value before a type check.
function litValue(v: unknown): unknown {
  if (v && typeof v === "object" && !Array.isArray(v) && "$" in v) return v.$;
  return v;
}

function str(v: unknown): string | null {
  const inner = litValue(v);
  return typeof inner === "string" ? inner : null;
}

function num(v: unknown): number | null {
  const inner = litValue(v);
  return typeof inner === "number" && Number.isFinite(inner) ? inner : null;
}

function bool(v: unknown): boolean | null {
  const inner = litValue(v);
  return typeof inner === "boolean" ? inner : null;
}

// An xsd:dateTime literal (or bare ISO string) -> epoch seconds, or null when
// it does not parse. Seconds, to match the epoch stamps the run is keyed by.
function epochSeconds(v: unknown): number | null {
  const s = str(v);
  if (s === null) return null;
  const ms = Date.parse(s);
  return Number.isNaN(ms) ? null : Math.round(ms / 1000);
}

// Parse a trainer-run PROV-JSON document into a flat derivation chain, or null
// when the document is unusable. Unusable means: not an object, or no
// mmly:TrainerRun activity carrying an integer version and two parseable
// timestamps — without the run there is no chain to anchor. Everything else is
// optional and rendered only if present.
export function parseProvChain(raw: unknown): ProvChain | null {
  if (!raw || typeof raw !== "object" || Array.isArray(raw)) return null;
  const doc = raw as Record<string, unknown>;

  const activities = recordSection(doc, "activity");
  const entities = recordSection(doc, "entity");
  const agents = recordSection(doc, "agent");

  // The run: the single mmly:TrainerRun activity. Its id encodes trained_at.
  let run: ProvRun | null = null;
  for (const [id, node] of Object.entries(activities)) {
    if (str(node["prov:type"]) !== "mmly:TrainerRun") continue;
    const m = /run\/(\d+)$/.exec(id);
    const trainedAt = m ? Number(m[1]) : NaN;
    const startedAt = epochSeconds(node["prov:startedAtTime"]);
    const endedAt = epochSeconds(node["prov:endedAtTime"]);
    if (!Number.isFinite(trainedAt) || startedAt === null || endedAt === null) {
      return null;
    }
    run = { trainedAt, startedAt, endedAt };
    break;
  }
  if (run === null) return null;

  // The software agent: read the first one grounded by a code sha.
  let agent: ProvAgent | null = null;
  for (const node of Object.values(agents)) {
    const codeSha = str(node["mmly:codeSha"]);
    if (codeSha === null) continue;
    agent = { codeSha, dirty: bool(node["mmly:dirty"]), producer: str(node["mmly:producer"]) };
    break;
  }

  // Inputs and artifacts, keyed by entity id so the derivation edges can point
  // back at them.
  let feed: ProvFeed | null = null;
  let feedId: string | null = null;
  let manifest: ProvManifest | null = null;
  let manifestId: string | null = null;
  const artifacts: Array<{ id: string; art: ProvArtifact }> = [];

  for (const [id, node] of Object.entries(entities)) {
    switch (str(node["prov:type"])) {
      case "mmly:GtfsStaticFeed": {
        const version = str(node["mmly:feedVersion"]);
        const sha256 = str(node["mmly:sha256"]);
        // A feed entity is grounded by its digest; without one it is not a real
        // feed node and we drop it rather than invent a version.
        if (version !== null && sha256 !== null) {
          feed = {
            version,
            sha256,
            start: str(node["mmly:feedStartDate"]),
            end: str(node["mmly:feedEndDate"]),
          };
          feedId = id;
        }
        break;
      }
      case "mmly:InputManifest": {
        const blake3 = str(node["mmly:blake3"]);
        const nAlertKeys = num(node["mmly:nAlertKeys"]);
        const nVehicleKeys = num(node["mmly:nVehicleKeys"]);
        const manifestVersion = num(node["mmly:manifestVersion"]);
        if (blake3 !== null && nAlertKeys !== null && nVehicleKeys !== null && manifestVersion !== null) {
          manifest = { blake3, nAlertKeys, nVehicleKeys, manifestVersion };
          manifestId = id;
        }
        break;
      }
      case "mmly:PublishedArtifact": {
        const bucketKey = str(node["mmly:bucketKey"]);
        if (bucketKey === null) break;
        artifacts.push({
          id,
          art: {
            name: str(node["mmly:name"]) ?? bucketKey,
            bucketKey,
            derivedFromFeed: false,
            derivedFromManifest: false,
          },
        });
        break;
      }
      default:
        break;
    }
  }

  // wasDerivedFrom edges: mark each artifact's real input(s). Only edges whose
  // used-endpoint is the feed or manifest we actually parsed count — a dangling
  // edge (the emitter cannot produce one) is ignored, never a guessed input.
  const derived = recordSection(doc, "wasDerivedFrom");
  for (const edge of Object.values(derived)) {
    const gen = str(edge["prov:generatedEntity"]);
    const used = str(edge["prov:usedEntity"]);
    if (gen === null || used === null) continue;
    const hit = artifacts.find((a) => a.id === gen);
    if (!hit) continue;
    if (feedId !== null && used === feedId) hit.art.derivedFromFeed = true;
    if (manifestId !== null && used === manifestId) hit.art.derivedFromManifest = true;
  }

  return { run, agent, feed, manifest, artifacts: artifacts.map((a) => a.art) };
}

// The three states a prov_ref fetch can land in, mirroring lib/feed.ts's
// TrainsFeed. There is deliberately no "loading" here — the caller owns that,
// before it has a URL to fetch. And there is no "absent": a snapshot with no
// prov_ref never calls this, because there is no document to point at.
export type ProvChainState =
  | { state: "unavailable"; reason: string }
  | { state: "malformed" }
  | { state: "ready"; chain: ProvChain };

// Fetch and parse a run's PROV-JSON document from its public URL
// (snapshot.provenance.prov_ref). A transport failure or a non-200 is
// `unavailable` (the network could not deliver the document); a body that is
// not JSON, or is JSON the reader cannot anchor to a run, is `malformed` (the
// document arrived but says nothing usable). The two are kept distinct so the
// page can label which kind of gap it hit. No cache option: the versioned
// document is immutable and long-cached upstream, so the browser's own cache
// policy is correct.
export async function fetchProvChain(url: string): Promise<ProvChainState> {
  let body: string;
  try {
    const res = await fetch(url);
    if (!res.ok) return { state: "unavailable", reason: `document returned ${res.status}` };
    body = await res.text();
  } catch (e) {
    return { state: "unavailable", reason: `not reachable (${(e as Error).message})` };
  }
  let raw: unknown;
  try {
    raw = JSON.parse(body);
  } catch {
    return { state: "malformed" };
  }
  const chain = parseProvChain(raw);
  return chain === null ? { state: "malformed" } : { state: "ready", chain };
}

File: training/prov.py
"""W3C PROV-JSON emitter for trainer-run provenance.

Every trainer run publishes a bundle of artifacts (params.json,
scheduled_headway.json, segment_params.json, the service-baseline sidecar) whose
ad-hoc `provenance`/`feed_version`/`training_corpus` blocks already answer, one
artifact at a time, "which code and which inputs produced this". This module
serializes that same lineage once, in a standard vocabulary — W3C PROV, in its
PROV-JSON encoding (https://www.w3.org/submissions/prov-json/) — so a consumer
can answer "what produced this, from what inputs" with a single grammar instead
of one bespoke reader per artifact.

The mapping this module emits:
  - prov:Entity      — the GTFS static feed (named by feed_version + a content
                       digest computed at fetch), the archive input manifest
                       (named by its BLAKE3 fingerprint), and each published
                       artifact (named by its immutable bucket key).
  - prov:Activity    — the trainer run itself, with startedAtTime/endedAtTime.
  - prov:SoftwareAgent — the trainer build, named by its code_sha (dirty flagged).
  - used              — run -> each input entity.
  - wasGeneratedBy    — each artifact entity -> run.
  - wasDerivedFrom    — an artifact -> the input entity it is built from (e.g.
                        scheduled_headway from the GTFS feed).
  - wasAssociatedWith — run -> agent.

GROUNDING RULE (the epic's hard constraint): this emitter emits ONLY relations
grounded in a recorded fact — a sha, a content digest, a BLAKE3 manifest hash, a
bucket key, or a timestamp. It is structurally impossible to emit an aspirational
edge: every entity/agent must be registered with its grounding fact before any
relation can reference it, and a relation whose endpoint was never registered
raises rather than emitting a dangling edge. So a run that could not fetch the
GTFS feed (no digest) simply carries no feed entity and no feed-derived edges,
instead of an ungrounded `used` claim.

Pure: no I/O, no clock, no network. Given fixed inputs the document serializes
byte-for-byte identically (canonical `to_json()` sorts keys and drops
whitespace), so the emitter is unit-testable against a golden fixture and the
published document is reproducible.
"""

from __future__ import annotations

import json
from dataclasses import dataclass, field
from datetime import UTC, datetime
from typing import Any

# Our own namespace for run-local identifiers and non-PROV attributes. The
# authority is nominal — it names the vocabulary, it is never dereferenced.
MMLY_NS = "https://momentarily.nyc/prov#"


class ProvenanceError(ValueError):
    """Raised on an attempt to emit an ungrounded or dangling PROV relation."""


def _xsd_datetime(epoch_seconds: int) -> dict[str, str]:
    """A PROV-JSON typed literal for an epoch second, as xsd:dateTime in UTC."""
    return {
        "$": datetime.fromtimestamp(epoch_seconds, UTC).isoformat(),
        "type": "xsd:dateTime",
    }


@dataclass
class ProvDocument:
    """An accumulating PROV-JSON document.

    Elements (entity/activity/agent) are registered with their grounding facts;
    relations reference elements by id and refuse to name an unregistered one.
    """

    _entity: dict[str, dict[str, Any]] = field(
        default_factory=dict[str, dict[str, Any]]
    )
    _activity: dict[str, dict[str, Any]] = field(
        default_factory=dict[str, dict[str, Any]]
    )
    _agent: dict[str, dict[str, Any]] = field(default_factory=dict[str, dict[str, Any]])
    _used: dict[str, dict[str, Any]] = field(default_factory=dict[str, dict[str, Any]])
    _generated: dict[str, dict[str, Any]] = field(
        default_factory=dict[str, dict[str, Any]]
    )
    _derived: dict[str, dict[str, Any]] = field(
        default_factory=dict[str, dict[str, Any]]
    )
    _associated: dict[str, dict[str, Any]] = field(
        default_factory=dict[str, dict[str, Any]]
    )

    def _known(self, node_id: str) -> bool:
        return (
            node_id in self._entity
            or node_id in self._activity
            or node_id in self._agent
        )

    def _require(self, *node_ids: str) -> None:
        for node_id in node_ids:
            if not self._known(node_id):
                raise ProvenanceError(
                    f"cannot relate unregistered node {node_id!r}: an entity, "
                    "activity, or agent must be added — with its grounding fact — "
                    "before a relation may reference it"
                )

    # --- elements ---------------------------------------------------------

    def entity(
        self, entity_id: str, *, ground: dict[str, Any], attributes: dict[str, Any]
    ) -> str:
        """Register an entity. `ground` is its non-empty recorded facts (a digest,
        a BLAKE3 hash, a bucket key); an empty `ground` is refused, which is what
        keeps an unnamed input out of the graph."""
        if not ground:
            raise ProvenanceError(
                f"entity {entity_id!r} has no grounding fact — an entity must be "
                "named by a recorded digest/hash/key or it does not get emitted"
            )
        self._entity[entity_id] = {**attributes, **ground}
        return entity_id

    def activity(
        self,
        activity_id: str,
        *,
        started_at: int,
        ended_at: int,
        attributes: dict[str, Any] | None = None,
    ) -> str:
        """Register an activity grounded by its start/end timestamps."""
        record: dict[str, Any] = {
            "prov:startedAtTime": _xsd_datetime(started_at),
            "prov:endedAtTime": _xsd_datetime(ended_at),
        }
        if attributes:
            record.update(attributes)
        self._activity[activity_id] = record
        return activity_id

    def agent(
        self, agent_id: str, *, ground: dict[str, Any], attributes: dict[str, Any]
    ) -> str:
        """Register an agent. `ground` is its recorded identity (the code_sha)."""
        if not ground:
            raise ProvenanceError(
                f"agent {agent_id!r} has no grounding fact — an agent must be "
                "named by its recorded code_sha"
            )
        self._agent[agent_id] = {**attributes, **ground}
        return agent_id

    # --- relations --------------------------------------------------------

    def used(self, activity_id: str, entity_id: str) -> None:
        """activity used entity. Both must already be registered."""
        self._require(activity_id, entity_id)
        rel_id = f"_:u{len(self._used) + 1}"
        self._used[rel_id] = {"prov:activity": activity_id, "prov:entity": entity_id}

    def was_generated_by(self, entity_id: str, activity_id: str) -> None:
        """entity wasGeneratedBy activity. Both must already be registered."""
        self._require(entity_id, activity_id)
        rel_id = f"_:g{len(self._generated) + 1}"
        self._generated[rel_id] = {
            "prov:entity": entity_id,
            "prov:activity": activity_id,
        }

    def was_derived_from(self, generated_id: str, used_id: str) -> None:
        """generated wasDerivedFrom used. Both must already be registered."""
        self._require(generated_id, used_id)
        rel_id = f"_:d{len(self._derived) + 1}"
        self._derived[rel_id] = {
            "prov:generatedEntity": generated_id,
            "prov:usedEntity": used_id,
        }

    def was_associated_with(self, activity_id: str, agent_id: str) -> None:
        """activity wasAssociatedWith agent. Both must already be registered."""
        self._require(activity_id, agent_id)
        rel_id = f"_:a{len(self._associated) + 1}"
        self._associated[rel_id] = {
            "prov:activity": activity_id,
            "prov:agent": agent_id,
        }

    # --- serialization ----------------------------------------------------

    def to_dict(self) -> dict[str, Any]:
        """The PROV-JSON document as a plain dict. Only non-empty relation maps
        are included, so a document with no derivations omits `wasDerivedFrom`
        rather than carrying an empty object."""
        doc: dict[str, Any] = {
            "prefix": {"mmly": MMLY_NS, "xsd": "http://www.w3.org/2001/XMLSchema#"},
            "entity": self._entity,
            "activity": self._activity,
            "agent": self._agent,
            "used": self._used,
            "wasGeneratedBy": self._generated,
            "wasAssociatedWith": self._associated,
        }
        if self._derived:
            doc["wasDerivedFrom"] = self._derived
        return doc

    def to_json(self) -> str:
        """Canonical, byte-stable serialization: sorted keys, no incidental
        whitespace. Identical inputs produce identical bytes."""
        return json.dumps(self.to_dict(), sort_keys=True, separators=(",", ":"))


# --- trainer-run assembler ------------------------------------------------
#
# Pure facts -> a trainer-run PROV document. Every argument is a recorded fact
# the trainer already holds; nothing here fetches or computes. The grounding
# rule lives in the shape of these dataclasses: a fact absent from the input is
# a node and edge absent from the output.


@dataclass(frozen=True)
class AgentFacts:
    """The trainer build that ran, from training.provenance.code_provenance."""

    code_sha: str
    dirty: bool | None
    producer: str


@dataclass(frozen=True)
class FeedFacts:
    """The GTFS static feed a run read, named by version AND a content digest
    computed over the fetched bytes at fetch time."""

    version: str
    sha256: str
    start: str | None = None  # feed_start_date, ISO
    end: str | None = None  # feed_end_date, ISO


@dataclass(frozen=True)
class ManifestFacts:
    """The archive input manifest a run fed on, named by its BLAKE3 fingerprint
    over the fetched object keys (alerts + vehicles)."""

    blake3: str
    n_alert_keys: int
    n_vehicle_keys: int
    manifest_version: int


@dataclass(frozen=True)
class ArtifactFacts:
    """One published artifact, named by its immutable versioned bucket key."""

    name: str  # short label, e.g. "params"
    bucket_key: str  # e.g. "state/params/v1788229972.json"
    derived_from_feed: bool = False  # built from the GTFS feed timetable
    derived_from_manifest: bool = False  # built from the archive manifest


def _entity_id(key: str) -> str:
    return f"mmly:{key}"


def build_trainer_run(
    *,
    trained_at: int,
    started_at: int,
    agent: AgentFacts,
    manifest: ManifestFacts,
    artifacts: list[ArtifactFacts],
    feed: FeedFacts | None = None,
) -> ProvDocument:
    """Assemble the PROV document for one trainer run.

    `feed` is optional: a run whose GTFS fetch failed carries no digest, so it
    gets no feed entity and no feed-derived edges — the grounding rule, applied
    by omission rather than by emitting an unnamed input.
    """
    doc = ProvDocument()

    run_id = f"mmly:run/{trained_at}"
    doc.activity(
        run_id,
        started_at=started_at,
        ended_at=trained_at,
        attributes={"prov:type": "mmly:TrainerRun"},
    )

    dirty_attr: dict[str, Any] = {}
    if agent.dirty is not None:
        dirty_attr["mmly:dirty"] = {"$": agent.dirty, "type": "xsd:boolean"}
    agent_id = f"mmly:trainer/{agent.code_sha}"
    doc.agent(
        agent_id,
        ground={"mmly:codeSha": agent.code_sha},
        attributes={
            "prov:type": "prov:SoftwareAgent",
            "mmly:producer": agent.producer,
            **dirty_attr,
        },
    )
    doc.was_associated_with(run_id, agent_id)

    manifest_id = _entity_id(f"manifest/{manifest.blake3}")
    doc.entity(
        manifest_id,
        ground={"mmly:blake3": manifest.blake3},
        attributes={
            "prov:type": "mmly:InputManifest",
            "mmly:manifestVersion": manifest.manifest_version,
            "mmly:nAlertKeys": manifest.n_alert_keys,
            "mmly:nVehicleKeys": manifest.n_vehicle_keys,
        },
    )
    doc.used(run_id, manifest_id)

    feed_id: str | None = None
    if feed is not None:
        feed_attrs: dict[str, Any] = {
            "prov:type": "mmly:GtfsStaticFeed",
            "mmly:feedVersion": feed.version,
        }
        if feed.start is not None:
            feed_attrs["mmly:feedStartDate"] = feed.start
        if feed.end is not None:
            feed_attrs["mmly:feedEndDate"] = feed.end
        feed_id = _entity_id(f"gtfs/{feed.sha256}")
        doc.entity(
            feed_id,
            ground={"mmly:sha256": feed.sha256},
            attributes=feed_attrs,
        )
        doc.used(run_id, feed_id)

    for art in artifacts:
        art_id = _entity_id(art.bucket_key)
        doc.entity(
            art_id,
            ground={"mmly:bucketKey": art.bucket_key},
            attributes={"prov:type": "mmly:PublishedArtifact", "mmly:name": art.name},
        )
        doc.was_generated_by(art_id, run_id)
        if art.derived_from_manifest:
            doc.was_derived_from(art_id, manifest_id)
        if art.derived_from_feed and feed_id is not None:
            doc.was_derived_from(art_id, feed_id)

    return doc
